central/complianceoperator/v2/profiles/datastore/datastore_impl.go

```diff
@@ -81,3 +85,115 @@ func (d *datastoreImpl) GetProfilesByClusters(ctx context.Context, clusterIDs []
 func (d *datastoreImpl) CountProfiles(ctx context.Context, q *v1.Query) (int, error) {
 	return d.searcher.Count(ctx, q)
 }
+
+type distinctProfileName struct {
+	ProfileName string `db:"compliance_profile_name"`
+}
+
+// GetProfilesNames gets the list of distinct profile names for the query
+func (d *datastoreImpl) GetProfilesNames(ctx context.Context, q *v1.Query, clusterIDs []string) ([]string, error) {
+	// Build the matching query to restrict profiles to the incoming clusters
+	readableClusterIDs := bestEffortClusters(ctx, clusterIDs)
+
+	var err error
+	q, err = withSACFilter(ctx, resources.Compliance, q)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedQuery := search.NewQueryBuilder().
+		AddExactMatches(search.ClusterID, readableClusterIDs...).
+		AddNumericField(search.ProfileCount, storage.Comparator_EQUALS, float32(len(readableClusterIDs))).
+		ProtoQuery()
+
+	parsedQuery = search.ConjunctionQuery(
+		parsedQuery,
+		q,
+	)
+
+	// Build the select and group by on distinct profile name
+	parsedQuery.Selects = []*v1.QuerySelect{
+		search.NewQuerySelect(search.ComplianceOperatorProfileName).Distinct().Proto(),
+	}
+	parsedQuery.GroupBy = &v1.QueryGroupBy{
+		Fields: []string{
+			search.ComplianceOperatorProfileName.String(),
+		},
+	}
+	parsedQuery.Pagination = q.GetPagination()
+
+	var results []*distinctProfileName
+	results, err = pgSearch.RunSelectRequestForSchema[distinctProfileName](ctx, d.db, schema.ComplianceOperatorProfileV2Schema, parsedQuery)
+	if err != nil {
+		return nil, err
+	}
+	if len(results) == 0 {
+		return nil, nil
+	}
+	profileNames := make([]string, 0, len(results))
+	for _, result := range results {
+		profileNames = append(profileNames, result.ProfileName)
+	}
+
+	return profileNames, err
+}
+
+type distinctProfileCount struct {
+	TotalCount int    `db:"compliance_profile_name_count"`
+	Name       string `db:"compliance_profile_name"`
+}
+
+// CountDistinctProfiles returns count of distinct profiles matching query
+func (d *datastoreImpl) CountDistinctProfiles(ctx context.Context, q *v1.Query, clusterIDs []string) (int, error) {
+	// Build the matching query to restrict profiles to the incoming clusters
+	readableClusterIDs := bestEffortClusters(ctx, clusterIDs)
+
+	query := search.ConjunctionQuery(
+		search.NewQueryBuilder().
+			AddExactMatches(search.ClusterID, readableClusterIDs...).
+			AddNumericField(search.ProfileCount, storage.Comparator_EQUALS, float32(len(readableClusterIDs))).ProtoQuery(),
+		q,
+	)
+
+	query.GroupBy = &v1.QueryGroupBy{
+		Fields: []string{
+			search.ComplianceOperatorProfileName.String(),
+		},
+	}
+
+	var results []*distinctProfileCount
+	results, err := pgSearch.RunSelectRequestForSchema[distinctProfileCount](ctx, d.db, schema.ComplianceOperatorProfileV2Schema, withCountQuery(query, search.ComplianceOperatorProfileName))
+	if err != nil {
+		return 0, err
+	}
+	return len(results), nil
+}
+
+func withCountQuery(q *v1.Query, field search.FieldLabel) *v1.Query {
```

Comment:
```suggestion
func withCountQuery(query *v1.Query, field search.FieldLabel) *v1.Query {
```

Reply:
Single-word variable names often confuse me because I think they could be a receiver in a Go struct.